plonky2x/src/frontend/vars/buffer.rs

```diff
@@ -1,27 +1,173 @@
+use core::marker::PhantomData;
+
+use plonky2::field::extension::Extendable;
+use plonky2::hash::hash_types::RichField;
+use plonky2::util::serialization::{IoResult, Read, Write};
+use serde::{Deserialize, Serialize};
+
 use super::{CircuitVariable, Variable};
-pub struct VariableBuffer<'a> {
-    variables: &'a [Variable],
-    position: usize,
+use crate::prelude::CircuitBuilder;
+
+#[derive(Debug, Clone)]
+pub struct ValueStream<F, const D: usize>(Stream<F>);
+
+#[derive(Debug, Clone)]
+pub struct VariableStream(Stream<Variable>);
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct OutputStream<F: RichField + Extendable<D>, const D: usize> {
+    hint_id: usize,
+    _marker: PhantomData<F>,
 }
 
-impl<'a> VariableBuffer<'a> {
-    pub fn new(variables: &'a [Variable]) -> Self {
+impl<F: RichField + Extendable<D>, const D: usize> OutputStream<F, D> {
+    pub(crate) fn new(hint_id: usize) -> Self {
         Self {
-            variables,
-            position: 0,
+            hint_id,
+            _marker: PhantomData,
+        }
+    }
+    pub fn read_exact(&self, builder: &mut CircuitBuilder<F, D>, len: usize) -> Vec<Variable> {
+        let variables = (0..len)
+            .map(|_| builder.init::<Variable>())
+            .collect::<Vec<_>>();
+        let stream = &mut builder
+            .hints
+            .get_mut(self.hint_id)
+            .expect("Hint not found")
+            .output_stream;
+        stream.0.write_slice(&variables);
+
+        variables
+    }
+    pub fn read<V: CircuitVariable>(&self, builder: &mut CircuitBuilder<F, D>) -> V {
+        let variables = self.read_exact(builder, V::nb_elements());
+        V::from_variables(&variables)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Stream<T> {
```

Comment:
If we have ValueStream, can we also have VariableStream instead of Stream<Variable>?